File: apps/backend/internal/store/osstate.go
package store

import (
	"context"
	"time"
)

// OsStateInterval is one closed device-state interval, clipped to the requested
// window so a stretch that straddles midnight is not counted twice across two
// days' reports.
type OsStateInterval struct {
	State     string `json:"state"`
	Ts        int64  `json:"ts"`
	DurationS int64  `json:"duration_s"`
}

// OsStateTotals is the time budget for a window. The four values are disjoint and
// together cover it exactly, which is the property that makes them trustworthy:
//
//	active + idle + suspended + offline == the elapsed part of the window
//
// Offline is not reported by the agent — a disconnected agent cannot report its
// own disconnection — so it is whatever the timeline does not account for.
type OsStateTotals struct {
	ActiveS    int64 `json:"active_s"`
	IdleS      int64 `json:"idle_s"`
	SuspendedS int64 `json:"suspended_s"`
	OfflineS   int64 `json:"offline_s"`
	// CoveredS is active+idle+suspended: the time the agent actually accounted
	// for. Exposed so the dashboard can show how complete a day's data is.
	CoveredS int64 `json:"covered_s"`
	// ElapsedS is the part of the window that has already happened.
	ElapsedS int64 `json:"elapsed_s"`
}

// OsStateReport is the device-state timeline plus its totals.
type OsStateReport struct {
	Totals        OsStateTotals     `json:"totals"`
	FirstActivity *int64            `json:"first_activity"`
	LastActivity  *int64            `json:"last_activity"`
	Intervals     []OsStateInterval `json:"intervals"`
}

// clippedDuration is the overlap between an interval and the window, in SQL.
// Both bounds are clamped, so an interval that starts before `from` or ends after
// `to` contributes only the part inside the window.
const clippedDuration = `LEAST(ts + duration_s, $4) - GREATEST(ts, $3)`

// The overlap predicate: any interval touching the window, not just ones starting
// inside it. A half-open comparison on `ts` alone would drop a long suspend that
// began the previous evening.
const overlapsWindow = `ts < $4 AND ts + duration_s > $3`

// OsStateReportFor returns the state timeline for one employee within [from, to).
// Scoped to businesses the caller owns, like every other per-employee read.
func (s *Store) OsStateReportFor(
	ctx context.Context, employeeID, ownerID string, from, to int64,
) (OsStateReport, error) {
	report := OsStateReport{Intervals: []OsStateInterval{}}

	owned, err := s.OwnsEmployee(ctx, ownerID, employeeID)
	if err != nil {
		return report, err
	}
	if !owned {
		return report, ErrForbidden
	}

	rows, err := s.pool.Query(ctx,
		`SELECT state, GREATEST(ts, $3) AS start_ts, `+clippedDuration+` AS duration_s
		   FROM os_states
		  WHERE user_id = $1 AND `+ownedFilter+` AND `+overlapsWindow+`
		  ORDER BY start_ts`,
		employeeID, ownerID, from, to)
	if err != nil {
		return report, err
	}
	defer rows.Close()

	for rows.Next() {
		var iv OsStateInterval
		if err := rows.Scan(&iv.State, &iv.Ts, &iv.DurationS); err != nil {
			return report, err
		}
		// A zero-width clip means the interval only touched the boundary.
		if iv.DurationS <= 0 {
			continue
		}
		report.Intervals = append(report.Intervals, iv)
		switch iv.State {
		case "active":
			report.Totals.ActiveS += iv.DurationS
		case "idle":
			report.Totals.IdleS += iv.DurationS
		case "suspended":
			report.Totals.SuspendedS += iv.DurationS
		}
	}
	if err := rows.Err(); err != nil {
		return report, err
	}

	// First/last moment the employee was actually at the machine, from the
	// clipped active intervals rather than from raw row timestamps.
	for _, iv := range report.Intervals {
		if iv.State != "active" {
			continue
		}
		start, end := iv.Ts, iv.Ts+iv.DurationS
		if report.FirstActivity == nil {
			first := start
			report.FirstActivity = &first
		}
		last := end
		report.LastActivity = &last
	}

	report.Totals.CoveredS = report.Totals.ActiveS + report.Totals.IdleS + report.Totals.SuspendedS

	// Only the part of the window that has already elapsed can be accounted for;
	// a range ending at 23:59 today must not report the rest of the day offline.
	now := time.Now().Unix()
	end := to
	if now < end {
		end = now
	}
	elapsed := end - from
	if elapsed < 0 {
		elapsed = 0
	}
	report.Totals.ElapsedS = elapsed

	offline := elapsed - report.Totals.CoveredS
	if offline < 0 {
		// Clock skew between device and server can make the timeline overshoot
		// the window. Report zero rather than a negative duration.
		offline = 0
	}
	report.Totals.OfflineS = offline

	return report, nil
}

File: apps/desktop/src-tauri/src/trackers/os_state.rs
//! Device state timeline (audit finding P0-2).
//!
//! `WindowTracker` records *which app had focus*, and only while the user was
//! active. That makes idle time the absence of rows, which is indistinguishable
//! from "the laptop was shut", "the agent was paused" and "the network was down" —
//! so idle time, total device time and first/last activity were not computable.
//!
//! This tracker records what the *machine* was doing as explicit, closed
//! intervals. The design guarantee that makes the numbers trustworthy:
//!
//!   **emitted intervals are contiguous and never overlap**, so for any window
//!   `sum(durations)` equals the wall-clock time the agent was running.
//!
//! Two deliberate differences from `WindowTracker`:
//!
//! * Durations come from wall-clock arithmetic (`now - start`), not from counting
//!   ticks. Counting assumes every poll is exactly one second, which it never is
//!   (the poll sleeps 1s and *then* does work), so tick-counting undercounts.
//! * A jump in the clock larger than the poll interval is treated as the machine
//!   having been suspended, and is emitted as its own interval rather than being
//!   silently absorbed into whatever was open.
//!
//! `offline` is not a state here: a disconnected agent cannot report that it is
//! disconnected. Offline is derived on the backend from gaps between intervals.
//! Screen lock is likewise absent until a platform probe exists for it — see
//! docs/FULL_SYSTEM_AUDIT.md §11 — rather than being guessed at.

/// A wall-clock gap larger than this means the process was not running: the
/// machine slept, was hibernated, or the agent was stopped. Well above the 1s
/// poll so ordinary scheduling jitter is never mistaken for a suspend.
pub const SUSPEND_GAP_S: i64 = 90;

/// Emit an in-progress interval at least this often, so a long idle stretch is
/// visible in the dashboard before it ends, and a crash loses at most this much.
pub const MAX_STATE_CHUNK_S: i64 = 300;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum OsState {
    Active,
    Idle,
    /// The process was not running: sleep, hibernate, or the agent being stopped.
    Suspended,
}

impl OsState {
    pub fn as_str(self) -> &'static str {
        match self {
            OsState::Active => "active",
            OsState::Idle => "idle",
            OsState::Suspended => "suspended",
        }
    }
}

/// One closed interval, ready to persist and sync.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct OsStateSample {
    pub state: OsState,
    pub start_ts: i64,
    pub duration_s: i64,
}

/// Pure decision core. `tick` is a function of (now, idle, threshold, collecting)
/// over the held interval, so the whole timeline is unit-testable without timers,
/// sleeping, or platform calls.
#[derive(Debug, Default)]
pub struct StateTracker {
    /// The open interval: (state, when it started).
    current: Option<(OsState, i64)>,
    /// When `tick` last ran, used to detect a suspend.
    last_tick: Option<i64>,
}

impl StateTracker {
    /// Advance one poll step. Returns every interval that closed on this step
    /// (usually none or one; a suspend can close two).
    ///
    /// * `now` — wall-clock unix seconds.
    /// * `idle_s` — seconds since the last user input.
    /// * `threshold_s` — idle seconds before the user counts as away.
    /// * `collecting` — whether capture is allowed right now (pause, schedule,
    ///   owner switch). While false the timeline stops rather than recording a
    ///   state nobody consented to.
    pub fn tick(
        &mut self,
        now: i64,
        idle_s: f64,
        threshold_s: i64,
        collecting: bool,
    ) -> Vec<OsStateSample> {
        let mut out = Vec::new();

        if !collecting {
            // Close what is open and stop the clock. `last_tick` is cleared so the
            // paused stretch is not later reported as a suspend.
            self.close_at(now, &mut out);
            self.last_tick = None;
            return out;
        }

        // A gap means this process was not running for that stretch.
        if let Some(last) = self.last_tick {
            if now.saturating_sub(last) > SUSPEND_GAP_S {
                self.close_at(last, &mut out);
                push(&mut out, OsState::Suspended, last, now);
                self.current = None;
            }
        }
        self.last_tick = Some(now);

        // A clock that moved backwards makes every duration meaningless. Restart
        // the timeline rather than emitting a negative or invented interval.
        if let Some((_, start)) = self.current {
            if now < start {
                self.current = None;
            }
        }

        let desired = if idle_s >= threshold_s as f64 {
            OsState::Idle
        } else {
            OsState::Active
        };

        match self.current {
            None => self.current = Some((desired, now)),

            Some((state, start)) if state == desired => {
                if now - start >= MAX_STATE_CHUNK_S {
                    push(&mut out, state, start, now);
                    self.current = Some((state, now));
                }
            }

            Some((state, start)) => {
                // The transition instant is not always "now". Going idle is only
                // *detected* after `threshold_s` of silence, but the user actually
                // stopped when that silence began — so the boundary is backdated.
                // Both intervals share the boundary, which keeps them contiguous
                // and stops the grace window being counted as active work.
                let boundary = match desired {
                    OsState::Idle => (now - threshold_s).max(start),
                    _ => now,
                };
                push(&mut out, state, start, boundary);
                self.current = Some((desired, boundary));
            }
        }

        out
    }

    /// Close the open interval at `at` (shutdown, pause, or a detected suspend).
    fn close_at(&mut self, at: i64, out: &mut Vec<OsStateSample>) {
        if let Some((state, start)) = self.current.take() {
            push(out, state, start, at);
        }
    }

    /// Close the open interval at `now`.
    ///
    /// Test-only. Production has no clean shutdown hook here: the tracker owns
    /// its state inside a detached loop, and reaching it from the tray's Quit
    /// handler would mean sharing it behind a lock for one write. It is not
    /// needed, because the stretch between a quit and the next start is caught by
    /// the suspend gap on the next boot — recorded as `suspended`, which is what
    /// it was: time the machine was not being monitored. The only cost is that a
    /// few final seconds land in `suspended` rather than `active`.
    #[cfg(test)]
    pub fn flush(&mut self, now: i64) -> Vec<OsStateSample> {
        let mut out = Vec::new();
        self.close_at(now, &mut out);
        self.last_tick = None;
        out
    }

    #[cfg(test)]
    fn open_state(&self) -> Option<OsState> {
        self.current.map(|(s, _)| s)
    }
}

/// Record `[start, end)`, dropping empty intervals so a zero-length transition
/// never reaches the database.
fn push(out: &mut Vec<OsStateSample>, state: OsState, start: i64, end: i64) {
    let duration_s = end - start;
    if duration_s > 0 {
        out.push(OsStateSample {
            state,
            start_ts: start,
            duration_s,
        });
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const THRESHOLD: i64 = 60;

    /// Drive the tracker one second at a time, as the real loop does.
    /// `idle_at` decides the idle reading for each second.
    fn run(
        tracker: &mut StateTracker,
        from: i64,
        to: i64,
        idle_at: impl Fn(i64) -> f64,
    ) -> Vec<OsStateSample> {
        let mut out = Vec::new();
        for now in from..to {
            out.extend(tracker.tick(now, idle_at(now), THRESHOLD, true));
        }
        out
    }

    fn total(samples: &[OsStateSample], state: OsState) -> i64 {
        samples
            .iter()
            .filter(|s| s.state == state)
            .map(|s| s.duration_s)
            .sum()
    }

    /// The invariant the whole feature rests on.
    fn assert_contiguous(samples: &[OsStateSample]) {
        for pair in samples.windows(2) {
            let (a, b) = (&pair[0], &pair[1]);
            assert_eq!(
                a.start_ts + a.duration_s,
                b.start_ts,
                "intervals must be contiguous: {a:?} then {b:?}"
            );
        }
        assert!(
            samples.iter().all(|s| s.duration_s > 0),
            "no empty intervals"
        );
    }

    #[test]
    fn active_hour_is_all_active_and_contiguous() {
        let mut t = StateTracker::default();
        let mut out = run(&mut t, 0, 3600, |_| 0.0);
        out.extend(t.flush(3600));

        assert_contiguous(&out);
        assert_eq!(total(&out, OsState::Active), 3600);
        assert_eq!(total(&out, OsState::Idle), 0);
    }

    #[test]
    fn idle_in_the_middle_splits_the_hour_without_gaps_or_overlap() {
        let mut t = StateTracker::default();
        // Active 0..1200, then no input until 2400, then active again.
        let mut out = run(&mut t, 0, 3600, |now| {
            // Input stops at 1200 and resumes at 2400.
            if (1200..2400).contains(&now) {
                (now - 1200) as f64
            } else {
                0.0
            }
        });
        out.extend(t.flush(3600));

        assert_contiguous(&out);
        // Every second of the hour is accounted for exactly once.
        assert_eq!(
            total(&out, OsState::Active) + total(&out, OsState::Idle),
            3600
        );
        // The idle stretch is real and is not swallowed by the active total.
        assert!(
            total(&out, OsState::Idle) > 1000,
            "idle: {}",
            total(&out, OsState::Idle)
        );
    }

    #[test]
    fn the_idle_grace_window_is_not_counted_as_active() {
        let mut t = StateTracker::default();
        // Input stops at t=100; idle grows from there and crosses the threshold
        // at t=160, when the tracker first notices.
        let mut out = run(&mut t, 0, 400, |now| {
            if now < 100 {
                0.0
            } else {
                (now - 100) as f64
            }
        });
        out.extend(t.flush(400));

        assert_contiguous(&out);
        let active = total(&out, OsState::Active);
        // Active must end when input stopped (100), not when idle was detected
        // (160) — otherwise every idle transition inflates active time by the
        // whole threshold.
        assert_eq!(active, 100, "active time leaked the grace window");
        assert_eq!(total(&out, OsState::Idle), 300);
    }

    #[test]
    fn a_clock_jump_is_recorded_as_suspended_not_absorbed() {
        let mut t = StateTracker::default();
        // Ticks land on 0..=99, so the last moment we observed the machine is 99.
        let last_live_tick = 99;
        let mut out = run(&mut t, 0, last_live_tick + 1, |_| 0.0);

        // The lid closes; the next tick lands eight hours later.
        let wake = 100 + 8 * 3600;
        out.extend(t.tick(wake, 0.0, THRESHOLD, true));
        out.extend(t.flush(wake + 10));

        assert_contiguous(&out);
        // The suspend covers exactly the unobserved stretch: last tick → wake.
        assert_eq!(total(&out, OsState::Suspended), wake - last_live_tick);
        // The active interval before the sleep is closed at the last live tick,
        // not stretched across the night.
        assert_eq!(total(&out, OsState::Active), last_live_tick + 10);
    }

    #[test]
    fn pausing_stops_the_timeline_and_does_not_backfill_a_suspend() {
        let mut t = StateTracker::default();
        let mut out = run(&mut t, 0, 100, |_| 0.0);

        // Capture turns off (schedule ends / owner disables monitoring).
        out.extend(t.tick(100, 0.0, THRESHOLD, false));
        assert_eq!(t.open_state(), None);

        // It comes back on hours later. That stretch is not ours to describe, so
        // it must not appear as suspended.
        let resume = 100 + 4 * 3600;
        out.extend(t.tick(resume, 0.0, THRESHOLD, true));
        out.extend(t.flush(resume + 60));

        assert_eq!(total(&out, OsState::Suspended), 0);
        assert_eq!(total(&out, OsState::Active), 100 + 60);
    }

    #[test]
    fn long_states_are_chunked_but_stay_contiguous() {
        let mut t = StateTracker::default();
        let mut out = run(&mut t, 0, 3600, |_| 0.0);
        out.extend(t.flush(3600));

        assert!(out.len() > 1, "a long state should be chunked");
        assert_contiguous(&out);
        assert!(out
            .iter()
            .all(|s| s.duration_s <= MAX_STATE_CHUNK_S && s.duration_s > 0));
    }

    #[test]
    fn a_backwards_clock_never_produces_a_negative_interval() {
        let mut t = StateTracker::default();
        let mut out = run(&mut t, 1000, 1100, |_| 0.0);
        // NTP corrects the clock backwards by an hour.
        out.extend(t.tick(1100 - 3600, 0.0, THRESHOLD, true));
        out.extend(t.flush(1100 - 3600 + 50));

        assert!(out.iter().all(|s| s.duration_s > 0));
    }

    #[test]
    fn brief_input_during_a_long_idle_does_not_fragment_the_timeline() {
        let mut t = StateTracker::default();
        // Idle all day except one keypress at t=1000.
        let mut out = run(&mut t, 0, 2000, |now| {
            if now < 1000 {
                now as f64
            } else {
                (now - 1000) as f64
            }
        });
        out.extend(t.flush(2000));

        assert_contiguous(&out);
        assert_eq!(
            total(&out, OsState::Active) + total(&out, OsState::Idle),
            2000
        );
    }
}
